core/trino-main/src/test/java/io/trino/operator/BenchmarkGroupByHashOnSimulatedData.java

```diff
@@ -0,0 +1,543 @@
+
+/**
+ * This class attempts to emulate aggregations done while running real-life queries.
+ * Some of the numbers here has been inspired by tpch benchmarks, however,
+ * there is no guarantee that results correlate with the benchmark itself.
+ */
+@State(Scope.Thread)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@Fork(3)
+@Warmup(iterations = 10, time = 500, timeUnit = TimeUnit.MILLISECONDS)
+@Measurement(iterations = 10, time = 500, timeUnit = TimeUnit.MILLISECONDS)
+@BenchmarkMode(Mode.AverageTime)
+public class BenchmarkGroupByHashOnSimulatedData
+{
+    private static final int DEFAULT_POSITIONS = 10_000_000;
+    private static final int EXPECTED_SIZE = 10_000;
+    private static final int DEFAULT_PAGE_SIZE = 8192;
+    private static final TypeOperators TYPE_OPERATORS = new TypeOperators();
+    private static final BlockTypeOperators TYPE_OPERATOR_FACTORY = new BlockTypeOperators(TYPE_OPERATORS);
+
+    @SuppressWarnings("MismatchedQueryAndUpdateOfCollection")
+    private final List<GroupByIdBlock> results = new ArrayList<>();
+
+    private final JoinCompiler joinCompiler = new JoinCompiler(TYPE_OPERATORS);
+
+    @Benchmark
+    @OperationsPerInvocation(DEFAULT_POSITIONS)
+    public Object groupBy(BenchmarkTpch data)
+    {
+        GroupByHash groupByHash =
+                GroupByHash.createGroupByHash(data.getTypes(), data.getChannels(), Optional.empty(), EXPECTED_SIZE, false, joinCompiler, TYPE_OPERATOR_FACTORY, NOOP);
+        addInputPages(groupByHash, data.getPages());
+
+        ImmutableList.Builder<Page> pages = ImmutableList.builder();
+        PageBuilder pageBuilder = new PageBuilder(groupByHash.getTypes());
+        for (int groupId = 0; groupId < groupByHash.getGroupCount(); groupId++) {
+            pageBuilder.declarePosition();
+            groupByHash.appendValuesTo(groupId, pageBuilder, 0);
+            if (pageBuilder.isFull()) {
+                pages.add(pageBuilder.build());
+                pageBuilder.reset();
+            }
+        }
+        pages.add(pageBuilder.build());
+        return pageBuilder.build();
+    }
+
+    @Test
+    public void runGroupByTpch()
+    {
+        BenchmarkGroupByHashOnSimulatedData benchmark = new BenchmarkGroupByHashOnSimulatedData();
+        for (double nullChance : new double[] {0, .1, .5, .9}) {
+            for (AggregationDefinition query : AggregationDefinition.values()) {
+                BenchmarkTpch data = new BenchmarkTpch(query, nullChance, 10_000);
+                data.setup();
+                benchmark.groupBy(data);
+            }
+        }
+    }
+
+    private void addInputPages(GroupByHash groupByHash, List<Page> pages)
+    {
+        results.clear();
+        for (Page page : pages) {
+            Work<GroupByIdBlock> work = groupByHash.getGroupIds(page);
+            boolean finished;
+            do {
+                finished = work.process();
+                results.add(work.getResult()); // Pretend the results are used
```

Comment:
use blackhole instead

Reply:
With `Blackhole` you cannot run it from outside the jmh context, so we cannot easily run test